types/module/module.go

```diff
@@ -518,10 +518,17 @@ func (m *Manager) InitGenesis(ctx sdk.Context, cdc codec.JSONCodec, genesisData
 		}
 	}
 
+	// TODO HV2: This check is not present in v1.
+	//  As staking is managed on L1,
+	//  and our stake module's InitGenesis does not return []abci.ValidatorUpdate
+	//  we should not need it. Is this assumption correct?
+
 	// a chain must initialize with a non-empty validator set
-	if len(validatorUpdates) == 0 {
-		return &abci.ResponseInitChain{}, fmt.Errorf("validator set is empty after InitGenesis, please ensure at least one validator is initialized with a delegation greater than or equal to the DefaultPowerReduction (%d)", sdk.DefaultPowerReduction)
-	}
+	/*
+		if len(validatorUpdates) == 0 {
+			return &abci.ResponseInitChain{}, fmt.Errorf("validator set is empty after InitGenesis, please ensure at least one validator is initialized with a delegation greater than or equal to the DefaultPowerReduction (%d)", sdk.DefaultPowerReduction)
+		}
+	*/
```

Comment:
I am still not fully convinced about this one. 
I need to rethink about our discussion and maybe even watch the video again.

My doubt is: if I re-enable this portion of the code, don't I need to enable the following lines in `app/export` function `ExportAppStateAndValidators` in heimdall-v2?

```
// TODO HV2: uncomment when implemented
// validators, err := staking.WriteValidators(ctx, app.StakingKeeper)
return servertypes.ExportedApp{
		AppState:        appState,
                 // TODO HV2: uncomment when implemented
		// Validators:      validators,
		Height:          height,
		ConsensusParams: app.BaseApp.GetConsensusParams(ctx),
	}, err
```

Meaning implementing the following function in our `v2` `stake` module, right?

```
// WriteValidators returns a slice of bonded genesis validators.
func WriteValidators(ctx sdk.Context, keeper *keeper.Keeper) (vals []cmttypes.GenesisValidator, returnErr error) {
	err := keeper.IterateLastValidators(ctx, func(_ int64, validator types.ValidatorI) (stop bool) {
		pk, err := validator.ConsPubKey()
		if err != nil {
			returnErr = err
			return true
		}
		cmtPk, err := cryptocodec.ToCmtPubKeyInterface(pk)
		if err != nil {
			returnErr = err
			return true
		}

		vals = append(vals, cmttypes.GenesisValidator{
			Address: sdk.ConsAddress(cmtPk.Address()).Bytes(),
			PubKey:  cmtPk,
			Power:   validator.GetConsensusPower(keeper.PowerReduction(ctx)),
			Name:    validator.GetMoniker(),
		})

		return false
	})
	if err != nil {
		return nil, err
	}

	return
}
```

This is also due to the fact that the `v2` function `InitGenesis` in `x/stake/keeper/genesis.go` does not return any `[]abci.ValidatorUpdate`, whilst this happens - for example - in cosmos' `staking` module, where `InitGenesis` also updates the `validatorSet` by invoking the function `ApplyAndReturnValidatorSetUpdates` at the very end.

Such function is instead called in `v2` during `stake` module's `EndBlocker`, which is responsible for the `validatorSet` update. However, the `validatorSet` update is apparently not performed in `v1` at `EndBlock`.
😕

Reply:
> My doubt is: if I re-enable this portion of the code, don't I need to enable the following lines in app/export function ExportAppStateAndValidators in heimdall-v2?

Do you mean populating the `Validators` field in `ExportedApp` ?
If yes, I checked around a bit. This is what I gathered (together with the code traversal we did during last week's meeting).

When you dump the state into a JSON ("export" command), what the SDK does is:
* dump the state of every module, including `staking`, by calling `simApp.ExportAppStateAndValidators`, which calls `ExportGenesis` on every module
* then, right at the end of `ExportCmd`, it "copies" the validator set from the output of `staking.WriteValidators`, into the genesis JSON **at the level of CometBFT** (do not confuse that call to `staking.WriteValidators` with the one in `InitGenesis`, they are two different callsites)
* however, as we saw last Thu, that field is just for "documentation" because, when you use the exported JSON as genesis to start a new chain (still in `simApp`), it will not use that information, but the state saved by the `staking` module which has all the information

So what will happen when the new chain is started is the following:
* CometBFT will read the validators fields as its level -- the one that had been populated with the output of `staking.WriteValidators` (in `ExportCmd`) -- and will pass it to `InitChain` in the `RequestInitChain` ABCI call.
* However `simApp` will then call `InitGenesis` on all modules, including `staking`, which will then return the list of validators -- accoding to its own saved state -- which will be shipped in `ResponseInitChain`, **and this will override those passed by CometBFT**.

Hmm, this is getting tricky for a GitHub comment. Feel free to reach out if you prefer a short meeting to discuss this.